src/main/java/com/google/devtools/build/lib/skyframe/RecursiveFilesystemTraversalValue.java

```diff
@@ -277,21 +277,31 @@ public static final class TraversalRequest implements SkyKey {
      */
     final boolean skipTestingForSubpackage;
 
-    /** Information to be attached to any error messages that may be reported. */
-    @Nullable final String errorInfo;
+    /**
+     * Whether to create nodes for directories, including empty ones.
+     */
+    final boolean emitDirectoryNodes;
+
+    /**
+     * Information to be attached to any error messages that may be reported.
+     */
+    @Nullable
+    final String errorInfo;
 
     private TraversalRequest(
         DirectTraversalRoot root,
         boolean isRootGenerated,
         PackageBoundaryMode crossPkgBoundaries,
         boolean strictOutputFiles,
         boolean skipTestingForSubpackage,
+        boolean emitDirectoryNodes,
```

Comment:
I think this may call for a test in `src/test/java/com/google/devtools/build/lib/skyframe/RecursiveFilesystemTraversalFunctionTest.java`.

Reply:
The recent refactorings made wiring this up much simpler - the change only affects `DirectoryArtifactTraversalRequest` now. Do you still think I should add a test there?

Reply:
The change affects the logic in `RecursiveFileSystemTraversalFunction`, I would still recommend adding a test case for that in `RecursiveFileSystemTraversalFunctionTest`